chalk-solve/src/infer/unify.rs

```diff
@@ -244,33 +271,116 @@ impl<'t, I: Interner> Unifier<'t, I> {
         {
```

Comment:
The above comment is not entirely correct with subtyping. It should be
```
for<'a...> T == for<'b...> U

if

for<'a...> exists<'b...> T :> U &&
for<'b...> exists<'a...> T <: U
```
and
```
for<'a...> T <: for<'b...> U

if

for<'b...> exists<'a...> T <: U
```

The code here should look like this

```rust
        if let Variance::Invariant | Variance::Contavariant = variance {
            let a_universal = self.table.instantiate_binders_universally(interner, a);
            let b_existential = self.table.instantiate_binders_existentially(interner, b);
            Zip::zip_with(self, Variance::Contavariant, &a_universal, &b_existential)?;
        }

        if let Variance::Invariant | Variance::Covariant = variance {
            let b_universal = self.table.instantiate_binders_universally(interner, b);
            let a_existential = self.table.instantiate_binders_existentially(interner, a);
            Zip::zip_with(self, Variance::Covariant, &a_existential, &b_universal)
        }
```